### ISSUE TITLE
turn wildcard imports into qualified imports

### ISSUE THREAD

**Issue by cvogt** [scalacenter/scalafix#182]
not sure how easy that is

**Comment by olafurpg** [comment]
This would be a really nice rewrite to have. However, it's a tricky one since we currently don't emit semantic information for sugars, see https://github.com/scalameta/scalameta/issues/822 Also, `Importee.Wildcard.symbol` is currently a `???`, see https://github.com/scalameta/scalameta/issues/925

Assuming we fix those issues, it should be possible to implement a `def usedSymbol(Importee.Wildcard): Seq[Symbol]` method that can be used to produce a patch. I'm happy to give more pointers if someone is interested.

**Comment by olafurpg** [comment]
Example diff produced by this rule

```diff
- import com.foo._
+ import com.foo.{A, B}
```
assuming that only `com.foo.{A,B}` are the only symbols used from the wildcard import. This is blocked by https://github.com/scalameta/scalameta/issues/1075

**Comment by gabro** [comment]
What would be the rewrite strategy for this?

**Comment by olafurpg** [comment]
We don't have the available data in semanticdb to resolve which symbols are imported from the wildcard.

**Comment by gabro** [comment]
And that would be #1075? Isn't that issue closed?

**Comment by olafurpg** [comment]
I just reopened the issue with a more detailed motivation

**Comment by olafurpg** [comment]
A quick update: synthetics now have attached symbols, which means scalafix knows symbols that are used by implicits. We also visit both macro expandees and expanders, however we could miss some used symbols in pathological cases, causing some symbols to be missing from the expanded wildcard import. I'm happy to give more pointers if someone is motivated to give this a try :)

**Comment by olafurpg** [comment]
This is a good idea. This should be significantly simpler now thanks to the new synthetics introduced in https://github.com/scalacenter/scalafix/pull/828. After the PR https://github.com/scalacenter/scalafix/pull/827, we are moving most built-in rules outside of the scalacenter/scalafix repository so that they can evolve at a different pace than Scalafix APIs. As long as a custom rule meets a certain level of quality, it can still 

- be automatically be included by sbt-scalafix
- have tab completion in the sbt shell like built-in rules
- be documented on the main Scalafix website

Ping us on Gitter if you believe your rule is good enough for inclusion. Don't hesitate to ask for help around code review or general guidance on writing/sharing custom rules.

**Comment by Jacke** [comment]
Any progress so far?

**Comment by vitalii-unzer** [comment]
Would be nice to expand wildcard imports into fully qualified imports.

**Comment by PawelLipski** [comment]
Hey @olafurpg... is it available as a ready rule in Scalafix already? or at least more easily doable? 🤔

**Comment by tgodzik** [comment]
I don't think this was ever worked on, it would require some time put into it.  Most blockers should be resolved aside from macro issue perhaps?

**Comment by bjaglin** [comment]
This clearly falls under the scope of `OrganizeImports`, which is now a built-in rule, so I am re-opening that. Not because there is active work planned on it, but because it should probably be done as a PR here if someone is willing to try. 

The behavior could honor https://scalacenter.github.io/scalafix/docs/rules/OrganizeImports.html#coalescetowildcardimportthreshold to avoid expanding too many explicit imports.
